### PR TITLE
feat(new sink)!: Transform existing 'tcp' sink into 'socket' sink. (#1212)

### PR DESCRIPTION
**Pull request by stbrody**
Signed-off-by: Spencer T Brody <spencer.t.brody@gmail.com>

### CODE REVIEW

File: src/sinks/socket.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+pub struct SocketSinkConfig {
+    #[serde(flatten)]
+    pub mode: Mode,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(tag = "mode", rename_all = "snake_case")]
+pub enum Mode {
+    Tcp(TcpSinkConfig),
+}
```
Comment (bruceg) [inline src/sinks/socket.rs]: I think it is worth writing an `impl From<TcpSinkConfig> for Mode`, which would eliminate the need for `Mode::Tcp(TcpSinkConfig)` elsewhere, and allow instead `TcpSinkConfig.into()`

File: src/sinks/socket.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        Self {
+            mode: Mode::Tcp(TcpSinkConfig::new(address)),
+        }
+    }
+}
+
+#[typetag::serde(name = "socket")]
+impl SinkConfig for SocketSinkConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        match self.mode.clone() {
```
Comment (bruceg) [inline src/sinks/socket.rs]: Using `self.mode.as_ref()` should work here to eliminate the need for `clone()`.

File: src/sinks/util/tcp.rs [hunk 3]
```diff
@@ -25,7 +26,7 @@ use tokio_tls::{Connect as TlsConnect, TlsConnector, TlsStream};
 use tracing::field;
 
 #[derive(Debug, Snafu)]
-enum BuildError {
+enum TcpBuildError {
```
[the remark this answers is not in the document]
Comment (stbrody) [reply src/sinks/util/tcp.rs]: This was conflicting with sinks::BuildError, which is also used in this file, causing a compile failure.  Instead I probably could have imported sinks::BuilderError as SinksBuildError or something, or just referred to it as sinks::BuildError throughout, but this seemed an equally reasonable and straightforward solution.  If you have a preference for one of the three please let me know!

File: src/sinks/socket.rs (hunk 1 above)
Comment (stbrody) [reply src/sinks/socket.rs]: Done!

File: src/sinks/socket.rs (hunk 2 above)
Comment (stbrody) [reply src/sinks/socket.rs]: as_ref() didn't work, but just adding an & at the beginning to make it a reference seemed to

File: src/sinks/util/tcp.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/util/tcp.rs]: That's fair enough. There is nothing magic about the name, just informing about the common naming. I had missed the `sinks::BuildError` import.

**Pull request merged**